### DISCUSSION TITLE
Why are page hrefs being lower-cased during auto-browsing?

### DISCUSSION THREAD

**Discussion by rbreesems** [cmu-sei/GHOSTS#100] [Q&A] [answered]
First off, nice work.  

I am new to Ghosts, and getting familiar with it by testing Ghosts  within a synthetic internet environment in a VM deployment using the latest binary installs. Our synthetic environment is set up such that if you point to a page and start clicking on links, you will never run out of pages, just goes around in circles.  I am just testing the browsing, and have tested with both Firefox and Chrome, and get the following results with both.

Our pages fairly simple, an image or two with text, followed by a list of links, where the 'hrefs' have full links in them, like:
 href="http:/<somedomain>/<somepage.html>" <some link text>.   I point Ghost at a page, and just let it follow links.

However, I find that Ghost is lowercasing all of the referenced URLs  before following a link.
So a link like somesite.com/S33.html is being transformed to somesite.com/s33.html and the link fails. 
The lowercased URL is present in the log/clientupdates file.

Is there someway to turn off this lowercasing of the URLs?

Thanks for your help.

**Comment by sei-dupdyke** [comment, marked as answer]
Hi, thanks for checking out the project. There is no current way to do this, but I can add a new version by the end of today, standby.

**Reply by sei-dupdyke** [reply]
The binaries have been updated to not lower-case URLs found within a page. Sorry for the confusion.

**Reply by rbreesems** [reply]
Sweet!  What a fast response!  Will check this out today, many thanks.

**Reply by rbreesems** [reply]
I have downloaded, installed the new Ghost Windows client.  This throws an exception 'Could not load file or assembly ... NLog, Version=5.0.0.0". The NLog dll in the ZIP archive is version 4.7+.  I tried manually updating NLog to version 5.0.0.0 but was unsuccessful. Would you mind updating the NLog version in your Ghosts Windows client ZIP archive?

BTW, I work with a government Cybersecurity team that deploys events for blue/red teaming, and we are evaluating Ghosts as a traffic generator to augment an existing traffic gen solution.

**Reply by sei-dupdyke** [reply]
hmm, wow that is weird, I didn't touch that and it still says 4.7 in the build file. Give me a bit to dig into this.

**Reply by rbreesems** [reply]
Will do.

**Reply by rbreesems** [reply]
This also may explain why when I manually updated to NLog 5.0.0.0, it threw a different exception saying that NLog 4.0.0.0 was missing.

**Reply by rbreesems** [reply]
Complete error below.
```
Unhandled Exception: System.TypeInitializationException: The type initializer for 'Ghosts.Domain.Code.ApplicationDetails' threw an exception. ---> System.IO.FileLoadException: Could not load file or assembly 'NLog, Version=5.0.0.0, Culture=neutral, PublicKeyToken=5120e14c03d0593c' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
   at Ghosts.Domain.Code.ApplicationDetails..cctor()
   --- End of inner exception stack trace ---
   at Ghosts.Domain.Code.ApplicationDetails.get_Name()
   at Ghosts.Client.Program.Main(String[] args) in C:\Projects\ghosts\src\Ghosts.Client\Program.cs:line 92

PS C:\ghosts>

**Reply by sei-dupdyke** [reply]
In the interest of time, I just updated binaries with a version using nlog5. I'll have to go back and see what happened with that previous build when I get a moment. Sorry for all the confusion.

**Reply by sei-dupdyke** [reply]
> BTW, I work with a government Cybersecurity team that deploys events for blue/red teaming, and we are evaluating Ghosts as a traffic generator to augment an existing traffic gen solution.

Email me if you want to connect further!

**Reply by rbreesems** [reply]
Yep, the new binaries with nlog5 works.  I will email you as I definitely want to connect further.

**Reply by rbreesems** [reply]
I don't see your email visible on your profile.
[my profile](https://github.com/rbreesems) has email - I would appreciate if you could shoot me an email so that I fill you in more on how your code is being used.
